opnieuw/test_util.py

```diff
@@ -0,0 +1,19 @@
+from contextlib import contextmanager
+from typing import Iterator, Optional
+
+from .retries import DoCall, RetryState, Action, __retry_state_namespaces
+
+
+class WaitLessRetryState(RetryState):
+    def __iter__(self) -> Iterator[Action]:
+        for _ in range(self.max_calls_total):
+            yield DoCall()
+
+
+@contextmanager
+def retry_immediately(namespace: Optional[str] = None) -> Iterator[None]:
+    __retry_state_namespaces[namespace] = WaitLessRetryState
+    try:
+        yield
+    finally:
+        __retry_state_namespaces[namespace] = RetryState
```

Comment:
This exit is not the inverse of the enter: if `__retry_state_namespaces['a']` was `Foo` before entering, then after leaving it will be set to `RetryState`. This can go wrong in the following case:

```python
with retry_immediately('a'):
    # Something in here will retry immediately.
    with retry_immediately('a'):
        # Something in here will retry immediately.
    # Something here will *not* retry immediately.
```

This example is of course artificial, but the same situation can arise in tests, when you use `retry_immediately`, and in that block, you call some other function, that internally uses `retry_immediately`.

The solution would be to store the old value before entering, and restore it later.

Reply:
Excellent point, thanks!